### PR TITLE
Update the list of fused_pairs and run validate_fused_group for specalization definitions too

### PR DESCRIPTION
**Pull request by vksnk**
Fixes https://github.com/halide/Halide/issues/6763.

From the issue above, the following program will produce incorrect code inside of the specializaiton branch:

```
Func out0, out1;
out0(x,y) = 1 * in(x,y);
out1(x,y) = 2 * in(x,y);

out0.vectorize(x, vs);
out1.vectorize(x, vs).compute_with(out0, x);

out0.specialize(in.dim(1).stride() == 128).fuse(x,y,f);
```

instead this should be an error just like in the case of the same code, but without specialization: 

```
Func out0, out1;
out0(x,y) = 1 * in(x,y);
out1(x,y) = 2 * in(x,y);

out0.vectorize(x, vs);
out1.vectorize(x, vs).compute_with(out0, x);

out0.fuse(x,y,f);
```
which will produce a following error:

```
terminate called after throwing an instance of 'Halide::CompileError'
  what():  Error: Invalid compute_with: cannot find x in out0.s0
```

After this PR the same error will be produced for the case with specialization branch.

### CODE REVIEW

File: src/RealizationOrder.cpp [hunk 1]
```diff
@@ -163,9 +163,15 @@ void populate_fused_pairs_list(const string &func, const Definition &def,
                    func, stage_index, fuse_level.var().name());
     if (fuse_level.stage_index() == 0) {
         parent.definition().schedule().fused_pairs().push_back(pair);
+        for (int i = 0; i < parent.definition().specializations().size(); i++) {
```
Comment (steven-johnson) [inline src/RealizationOrder.cpp:166]: style nit: `for (const auto s : parent.definition().specializations())` might be a better choice here

File: src/ScheduleFunctions.cpp [hunk 2]
```diff
@@ -2369,9 +2369,17 @@ void validate_fused_groups_schedule(const vector<vector<string>> &fused_groups,
 
             validate_fused_group_schedule_helper(
                 iter->first, 0, iter->second.definition(), env);
+            for (size_t i = 0; i < iter->second.definition().specializations().size(); i++) {
```
Comment (steven-johnson) [inline src/ScheduleFunctions.cpp:2372]: See similar nits above

**Review by steven-johnson — APPROVED** [review]
LGTM with nits

File: src/RealizationOrder.cpp (hunk 1 above)
Comment (vksnk) [reply src/RealizationOrder.cpp]: Done

File: src/RealizationOrder.cpp [hunk 3]
```diff
@@ -163,9 +163,15 @@ void populate_fused_pairs_list(const string &func, const Definition &def,
...
+            parent.definition().specializations()[i].definition.schedule().fused_pairs().push_back(pair);
+        }
     } else {
         internal_assert(fuse_level.stage_index() > 0);
         parent.update(fuse_level.stage_index() - 1).schedule().fused_pairs().push_back(pair);
+        for (int i = 0; i < parent.update(fuse_level.stage_index() - 1).specializations().size(); i++) {
```
[the remark this answers is not in the document]
Comment (vksnk) [reply src/RealizationOrder.cpp]: Done

File: src/ScheduleFunctions.cpp (hunk 2 above)
Comment (vksnk) [reply src/ScheduleFunctions.cpp]: Done

**Pull request merged**